File: server/src/test/kotlin/com/splitme/core/DebtSimplifierTest.kt
package com.splitme.core

import com.splitme.model.TransferRoute
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class DebtSimplifierTest {

    @Test
    fun `test simple greedy matching with 2 debtors and 1 creditor`() {
        val netBalances = mapOf(
            "Alice" to -100.0,
            "Bob" to -50.0,
            "Charlie" to 150.0
        )

        val transfers = DebtSimplifier.simplifyDebts(netBalances)

        assertEquals(2, transfers.size)
        val totalTransferred = transfers.sumOf { it.amount }
        assertEquals(150.0, totalTransferred)

        assertTrue(transfers.any { it.fromMemberId == "Alice" && it.toMemberId == "Charlie" && it.amount == 100.0 })
        assertTrue(transfers.any { it.fromMemberId == "Bob" && it.toMemberId == "Charlie" && it.amount == 50.0 })
    }

    @Test
    fun `test exact match optimization prioritizes pairs of equal debt and credit`() {
        val netBalances = mapOf(
            "A" to -200.0,
            "B" to -300.0,
            "C" to 200.0,
            "D" to 300.0
        )

        val transfers = DebtSimplifier.simplifyDebts(netBalances)

        assertEquals(2, transfers.size)
        assertTrue(transfers.any { it.fromMemberId == "A" && it.toMemberId == "C" && it.amount == 200.0 })
        assertTrue(transfers.any { it.fromMemberId == "B" && it.toMemberId == "D" && it.amount == 300.0 })
    }

    @Test
    fun `test cyclic debt resolution collapses to zero transfers when net balances are zero`() {
        val netBalances = mapOf(
            "A" to 0.0,
            "B" to 0.0,
            "C" to 0.0,
            "D" to 0.0
        )

        val transfers = DebtSimplifier.simplifyDebts(netBalances)

        assertTrue(transfers.isEmpty())
    }

    @Test
    fun `test complex multi-party debt simplifies to at most N minus 1 transfers`() {
        val netBalances = mapOf(
            "m1" to -500.0,
            "m2" to -300.0,
            "m3" to -200.0,
            "m4" to -100.0,
            "m5" to 400.0,
            "m6" to 700.0
        )
        // 6 non-zero members -> transfers should be <= 5
        val transfers = DebtSimplifier.simplifyDebts(netBalances)

        assertTrue(transfers.size <= 5, "Transfers size ${transfers.size} should be <= 5")

        // Verify all debts are completely settled
        val remaining = netBalances.toMutableMap()
        for (t in transfers) {
            remaining[t.fromMemberId] = (remaining[t.fromMemberId] ?: 0.0) + t.amount
            remaining[t.toMemberId] = (remaining[t.toMemberId] ?: 0.0) - t.amount
        }

        for ((member, bal) in remaining) {
            assertEquals(0.0, bal, 0.001, "Member $member balance not settled: $bal")
        }
    }

    @Test
    fun `test empty balance map returns empty transfers`() {
        val transfers = DebtSimplifier.simplifyDebts(emptyMap())
        assertTrue(transfers.isEmpty())
    }
}

File: client/src/core/__tests__/debtSimplifier.test.ts
import { describe, it, expect } from 'vitest';
import { simplifyDebts } from '../debtSimplifier';

describe('DebtSimplifier', () => {
  it('should simplify basic debt matching with 2 debtors and 1 creditor', () => {
    const netBalances: Record<string, number> = {
      Alice: -100,
      Bob: -50,
      Charlie: 150,
    };

    const transfers = simplifyDebts(netBalances);

    expect(transfers).toHaveLength(2);
    const totalTransferred = transfers.reduce((acc, t) => acc + t.amount, 0);
    expect(totalTransferred).toBe(150);

    expect(transfers).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ fromMemberId: 'Alice', toMemberId: 'Charlie', amount: 100 }),
        expect.objectContaining({ fromMemberId: 'Bob', toMemberId: 'Charlie', amount: 50 }),
      ])
    );
  });

  it('should prioritize exact matches of equal debt and credit', () => {
    const netBalances: Record<string, number> = {
      A: -200,
      B: -300,
      C: 200,
      D: 300,
    };

    const transfers = simplifyDebts(netBalances);

    expect(transfers).toHaveLength(2);
    expect(transfers).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ fromMemberId: 'A', toMemberId: 'C', amount: 200 }),
        expect.objectContaining({ fromMemberId: 'B', toMemberId: 'D', amount: 300 }),
      ])
    );
  });

  it('should return empty array for all-zero net balances', () => {
    const netBalances: Record<string, number> = {
      A: 0,
      B: 0,
      C: 0,
    };

    const transfers = simplifyDebts(netBalances);
    expect(transfers).toEqual([]);
  });

  it('should bound transfer count to at most N - 1 for complex party network', () => {
    const netBalances: Record<string, number> = {
      m1: -500,
      m2: -300,
      m3: -200,
      m4: -100,
      m5: 400,
      m6: 700,
    };

    const transfers = simplifyDebts(netBalances);
    expect(transfers.length).toBeLessThanOrEqual(5);

    const remaining = { ...netBalances };
    for (const t of transfers) {
      remaining[t.fromMemberId] += t.amount;
      remaining[t.toMemberId] -= t.amount;
    }

    for (const [, bal] of Object.entries(remaining)) {
      expect(Math.abs(bal)).toBeLessThan(0.001);
    }
  });

  it('should handle empty map', () => {
    expect(simplifyDebts({})).toEqual([]);
  });
});
